src/app/pages/system/general-settings/localization-form2/localization-form2.component.ts

```diff
@@ -0,0 +1,179 @@
+    tooltip: helptext.stg_kbdmap.tooltip,
+  };
+
+  timezone: {
+    readonly fcName: 'timezone';
+    label: string;
+    tooltip: string;
+    options?: Observable<Option[]>;
+  } = {
+    fcName: 'timezone',
+    label: helptext.stg_timezone.placeholder,
+    tooltip: helptext.stg_timezone.tooltip,
+  };
+
+  dateFormat: {
+    readonly fcName: 'date_format';
+    label: string;
+    tooltip: string;
+    options?: Observable<Option[]>;
+  } = {
+    fcName: 'date_format',
+    label: helptext.date_format.placeholder,
+    tooltip: helptext.date_format.tooltip,
+  };
+
+  timeFormat: {
+    readonly fcName: 'time_format';
+    label: string;
+    tooltip: string;
+    options?: Observable<Option[]>;
+  } = {
+    fcName: 'time_format',
+    label: helptext.time_format.placeholder,
+    tooltip: helptext.time_format.tooltip,
+  };
+
+  private configData: SystemGeneralConfig;
+
+  constructor(
+    private sysGeneralService: SystemGeneralService,
+    private fb: FormBuilder,
+    public localeService: LocaleService,
+    protected ws: WebSocketService,
+    protected langService: LanguageService,
+    private modalService: ModalService,
+    private router: Router,
+  ) {
+    this.sysGeneralService.getGeneralConfig$
+      .pipe(untilDestroyed(this)).subscribe((res) => {
+        this.configData = res;
+        this.formGroup?.get('kbdmap').setValue(this.configData.kbdmap);
+        this.formGroup?.get('language').setValue(this.configData.language);
+        this.formGroup?.get('timezone').setValue(this.configData.timezone);
+        this.setTimeOptions(this.configData.timezone);
+      });
+  }
+
+  ngOnInit(): void {
+    this.formGroup = this.fb.group({
+      language: [this.configData?.language],
+      kbdmap: [this.configData?.kbdmap],
+      timezone: [this.configData?.timezone],
+      date_format: [this.localeService.getPreferredDateFormat()],
+      time_format: [this.localeService.getPreferredTimeFormat()],
+    });
+
+    this.makeLanguageList();
+    this.makeKbdMapOptions();
+    this.makeTimezoneOptions();
+    this.setTimeOptions(this.configData?.timezone);
+  }
+
+  makeLanguageList(): void {
+    this.sysGeneralService.languageChoices().pipe(untilDestroyed(this)).subscribe((languageList) => {
+      const options = Object.keys(languageList || {}).map((key) => ({
+        label: this.sortLanguagesByName
+          ? `${languageList[key]} (${key})`
+          : `${key} (${languageList[key]})`,
+        value: key,
+      }));
+      this.language.options = of(_.sortBy(
+        options,
+        this.sortLanguagesByName ? 'label' : 'value',
+      ));
+      this.formGroup.get('language').setValue(this.configData?.language);
+    });
+  }
+
+  makeKbdMapOptions(): void {
+    this.sysGeneralService.kbdMapChoices().pipe(untilDestroyed(this)).subscribe((mapChoices) => {
+      this.kbdMap.options = of(mapChoices);
+      this.formGroup.get('kbdmap').setValue(this.configData?.kbdmap);
```

Comment:
This form is complicated. There are two cases for `configData` to get assigned a value. One is that the value gets passed to it from the parent component. Which will be almost instant since it involves no middleware call. In which case, this line will be used to assign value and the other place might be too soon since the form hasn't initialized yet.
The other case is where the `configData` is request from middleware and takes a while. In which case, this will be too soon since `configData` isn't available yet.

Reply:
How is it being passed from the parent component?

Reply:
Basically, for now `getGeneralConfig$` observer is used which gets values from middleware. But when a slide-in panel is implemented, the form will be receiving config through its parent component which will be through `sendGeneralConfig$` and in that case this will be the correct code.